Fix playing skinned samples being unsafe during skin changes

peppy opened the pull request:
Aims to close #5798.

- [x] Depends on ppy/osu-framework#4203.

PRing for discussion. I was hoping this would be a simple local change but in order to embed the reference the interface change seems to be required. Does seem to solve this issue, although worth noting that we now need a method of stopping previous `DrawableSample`s from playing (looping sounds will now get stuck in a playing state on skin change because there's nothing stopping them).

peppy committed:
Use ISample everywhere in Skin GetSample lookup path

peppy committed:
Move dispose method to end of file

peppy committed:
Add sample wrapper in LegacySkin to keep a reference and avoid GC death

peppy committed:
Update interface implementations with framework changes

smoogipoo committed:
Merge branch 'master' into fix-unsafe-skinnable-sample-play

smoogipoo committed:
Implement some new methods

smoogipoo committed:
Move nested class to bottom of file

peppy commented:
Uh, don't merge this yet. This breaks looping behaviour as mentioned in the OP. Needs further consideration/discussion.

smoogipoo commented:
Are you sure about that? On a skin change, the `DrawableSample`'s of `PoolableSkinnableSample` get cleared / disposed, which disposes the underlying `Sample` which disposes `SampleChannels` which stops them.

peppy commented:
I could reproduce in game, I haven't checked the code though. Just change skins using the random skin key rapidly during sliders.

bdach committed:
Merge branch 'master' into fix-unsafe-skinnable-sample-play

bdach committed:
Fix invalid xmldoc indent

bdach commented:
Since there have been more reports than usual of this recently I've given this a nudge (in the form of a merge, and fixing a CI issue), but there's still something wrong with stopping looped samples indeed. A good way to reproduce is to enter autoplay on a map with a spinner (so that `spinnerspin` plays) and then quitting out to menu; the sample will continue to play.

I have experimented lightly with stopping the sample in `PoolableSkinnableSample.FreeAfterUse()`, but it had no effect whatsoever. Not sure where to take it from here, will dig around more and maybe come up with something.

bdach commented:
Doing this seems to help:

```diff
diff --git a/osu.Game/Skinning/SkinnableSound.cs b/osu.Game/Skinning/SkinnableSound.cs
index edd3a2cdd3..aca5dd6418 100644
--- a/osu.Game/Skinning/SkinnableSound.cs
+++ b/osu.Game/Skinning/SkinnableSound.cs
@@ -166,6 +166,14 @@ private void updateSamples()
                 Play();
         }
 
+        protected override void Dispose(bool isDisposing)
+        {
+            if (isDisposing)
+                Stop();
+
+            base.Dispose(isDisposing);
+        }
+
         #region Re-expose AudioContainer
 
         public BindableNumber<double> Volume => samplesContainer.Volume;
```

(The `isDisposing` guard is there only because I'm scared of that `Stop()` running in a finaliser.)

I think it's possibly defensible semantics-wise to do this, but it still feels like quite a heavy-handed change just for this - and it relies a bit on the disposal queue to do its thing...

peppy commented:
I believe that change would help on exiting gameplay, but the issue is deeper than that - looping sounds will continue looping after a skin change *without* exiting gameplay. The best solution may be to disable the loop flag at some point along the way.

I'm going to take another look into this today, hopefully figuring out the above.

peppy committed:
Remove unnecessary duplicated skin changed handling

For some reason we were handling this both in `DrawableSkinnableSound`
and `PoolableSkinnableSample` in very similar ways. Only one seems
required.

peppy committed:
Ensure looping is disabled on old samples when switching skins

peppy commented:
This should be good to go now. Simply a matter of disabling looping when switching away to a new skin.

peppy committed:
Fix regression meaning `SkinnableSound` initialisation may never happen

smoogipoo commented on osu.Game/Skinning/PoolableSkinnableSample.cs:
```diff
@@ -83,6 +83,14 @@ private void updateSample()
+                // when switching away from previous samples, we don't want to call Stop() on them as it sounds better to let them play out.
+                // this may change in the future if we use PoolableSkinSample in more locations than gameplay.
+                // we *do* want to turn off looping, else we end up with an infinite looping sample running in the background.
+                activeChannel.Looping = false;
```
This is suspicious.

The `sampleContainer.Clear()` line below should dispose the `DrawableSample`, which should dispose the `Sample` (because the ctor param `disposeSampleOnDisposal` is left as default), which should dispose all contained sample channels, which stops them from playing (https://github.com/ppy/osu-framework/blob/bc9cc38a2513e99b38b5ade98461537fd341cab0/osu.Framework/Audio/Sample/SampleChannel.cs#L29-L35).

What am I missing here?

peppy replied:
Looks like the `LegacySkinSample` I implemented doesn't implement `IDisposable`, which means the sample it's encapsulating doesn't actually get the disposal flow fired. Let me try fixing that....

peppy replied:
Works, have amended. Good catch.

peppy committed:
Rename incorrect variable

peppy committed:
Merge branch 'master' into fix-unsafe-skinnable-sample-play

peppy committed:
Fix disposal rather than performing some weird hack

bdach commented on osu.Game/Skinning/SkinnableSound.cs:
```diff
@@ -131,6 +131,14 @@ public virtual void Play()
 
+        protected override void LoadAsyncComplete()
+        {
+            base.LoadAsyncComplete();
+
+            if (!samplesContainer.Any())
+                updateSamples();
```
CI (namely storyboard tests) are failing on this in particular. `DrawableStoryboardSample` is doing sample rate application using active mods in `SkinChanged()`, which is in turn called on the sample by the `base.LoadAsyncComplete()` call - but at that point the samples don't exist yet (`updateSamples()` will create them immediately after), therefore the attempt to set the sample rate will fail.

Reordering this as such fixes:

```suggestion
			// ensure samples are constructed before calling SkinChanged()
			// via base.LoadAsyncComplete().
            if (!samplesContainer.Any())
                updateSamples();
            
            base.LoadAsyncComplete();
```

but I'm not super sure this is an acceptable resolution.

peppy replied:
I think this is fine with the local comment for now.

Not actually sure it needs to be in `LoadAsyncComplete` in the first place, but that's for another day's investigation.

peppy committed:
Fix back to front initialisation order
